File: c++/Solutions/basic_calculator.cpp
#include <stack>
#include <numeric>
#include <vector>
#include <algorithm>

#include "basic_calculator.hpp"

namespace basic_calculator
{
int Solution::basic_calculator(std::string pattern)
{
    std::stack<char> expre;
    std::string pattern_no_space = strReplace(pattern," ","");
    std::string pattern_no_parenthes = "";
    long long result = 0;
   
    for(int i = 0; i < pattern_no_space.length(); i++)
    {
        if (pattern_no_space[i] != ')')
        {
            expre.push(pattern_no_space[i]);
        }
        else
        { 
            std::string subPattern = "";
            while(true)
            {
                if (expre.top() != '(')
                {
                    // get content in a (), this is a true loop
                    subPattern += expre.top();
                    expre.pop();
                }
                else
                {
                    // calculate a (),
                    // then load the result and rest pattern_no_space into string expre
                    expre.pop();
                    std::reverse(subPattern.begin(), subPattern.end());
                    auto longtempRes = calculator_no_parenthes(subPattern);
                    if (expre.empty())
                    {
                        if (i == (pattern_no_space.length() -1))
                        {
                            return longtempRes;
                        }
                        else
                        {
                           return longtempRes + basic_calculator(pattern_no_space.substr(i+1,pattern_no_space.length()-1));
                        }
                        
                    }
                    auto tempRes = std::to_string(longtempRes);
                    int j = 0;
                    if ('-' == tempRes[0])
                    {
                        j = 1;
                        if ('-' == expre.top())
                        {
                            expre.top() = '+';
                        }
                        else
                        {
                            expre.top() = '-';
                        }
                        
                    }
                    for(j; j < tempRes.length(); j++)
                    {
                        expre.push(tempRes[j]);
                    }
                    break;
                }
               
            }
        }
        
    }
    while(!expre.empty())
    {
        pattern_no_parenthes += expre.top();
        expre.pop();
    }
    std::reverse(pattern_no_parenthes.begin(), pattern_no_parenthes.end());
    return calculator_no_parenthes(pattern_no_parenthes);
}

long long Solution::calculator_no_parenthes(std::string pattern)
{
    int len = pattern.length();
    long long result = 0;
    std::vector<long long> numbers;
    int sign = 1;
    if (pattern[0] == '-')
    {
        sign = -1;
    }
    long long temp = 0;
    for (int i = 0; i < len; i++)
    {   
        if (pattern[i] >=  '0' && pattern[i] <= '9')
        {
            temp = temp * 10 + pattern[i] - '0';
            if (i == len -1)
            {
                numbers.push_back(temp * sign);
            }
        }
        else
        {
            numbers.push_back(temp * sign);
            temp = 0;
            if (pattern[i] == '-')
            {
                sign = -1;
            }
            else
            {
                sign = 1;
            }
            
        }  
    }
    return std::accumulate(numbers.begin(),numbers.end(),0);
}

std::string Solution::strReplace(std::string sentence,
    std::string src,
    std::string dst)
{
    int pos = 0;
    while ((pos = sentence.find(src, pos)) != std::string::npos)
    {
        sentence.replace(pos, src.length(), dst);
        pos = pos + dst.length();
    }
    return sentence;
}
}

File: python/solution/basic_calculator.py
from typing import List
import string
class Solution:
    def basic_calculator(self, pattern:str) -> int:
        """
        :type pattern: string
        :rtype: float
        """
        pattern_no_space = pattern.replace(" ","")
        pattern_no_parenthes = ""
        expre_stack = []
        for i in range(0,len(pattern_no_space)):
            if pattern_no_space[i] != ')':
                expre_stack.append(pattern_no_space[i])
            else:
                sign = 1
                subpattern = ""
                while sign == 1:
                    if expre_stack[-1] != "(":
                        subpattern += expre_stack.pop()
                    else:
                        expre_stack.pop()
                        reveres_subpattern = subpattern[::-1]
                        tempRes = self.__calculator_no_parenthes(reveres_subpattern)
                        if len(expre_stack) == 0:
                            if i == len(pattern_no_space) - 1:
                                return tempRes
                            else:
                                return tempRes + self.basic_calculator(pattern_no_space[i+1:])
                        else:
                            j = 0
                            str_tempRes = str(tempRes)
                            if '-' == str_tempRes[0]:
                                j = 1
                                if '-' == expre_stack[-1]:
                                    expre_stack[-1] = '+'
                                else:
                                    expre_stack[-1] = '-'
                            for i in range(j, len(str_tempRes)):
                                expre_stack.append(str_tempRes[i])
                            sign = 0

        while 0 != len(expre_stack):
            pattern_no_parenthes += expre_stack.pop()
        
        return self.__calculator_no_parenthes(pattern_no_parenthes[::-1])
    
    def __calculator_no_parenthes(self, pattern:str) -> int:
        print("__calculator_no_parenthes str is %s \n"%(pattern))
        length = len(pattern)
        sign = 1
        numbers = []
        if pattern[0] == '-':
            sign = -1
        
        temp = 0
        for i in range(0,length):
            if pattern[i] >= '0' and pattern[i] <= '9':
                temp = temp * 10 + int(pattern[i])
                if length - 1 == i:
                    numbers.append(temp * sign)
            else:
                numbers.append(temp * sign)
                temp = 0
                if pattern[i] == '-':
                    sign = -1
                else:
                    sign = 1
        return sum(numbers)
